### PR TITLE
Close cursor if it's opened by different tx

### PR DESCRIPTION
**Pull request by awskii**
So we keep cursor interface opened inside `DomainRoTx`. If we then create new tx and try to read with it from `DomainRoTx`, value will be fetched via previously opened cursor which is not expected behaviour.

I set stupid fix for that but i assume there could be better solutions.

### CODE REVIEW

**Comment by AskAlexSharov** [comment]
@awskii maybe we must `panic` in this case? just to enforce next invariant: 
- `temporal.Tx` guarantee that lifetime of dbtx is longer than lifteme of agate
- not allowed to move `doms` between goroutines

**Comment by awskii** [comment]
@AskAlexSharov from one side yes, it would be better to panic with info, but i had a small intention to reuse sharedDomains without closing and reopening (requires only to set up new aggtx) but this could end up badly.

**Comment by AskAlexSharov** [comment]
`sharedDomains without closing and reopening (requires only to set up new aggtx)` where do you need it?

**Comment by awskii** [comment]
i found it useful while did commitment rebuilding. I am going to add panic as you said.

**Review by AskAlexSharov — APPROVED** [review]

File: erigon-lib/state/domain.go [hunk 1]
```diff
@@ -1701,10 +1705,39 @@ func (dt *DomainRoTx) statelessBtree(i int) *BtIndex {
...
+		dt.valsC.Close()
+		dt.vcParentPtr.Store(0)
+		dt.valsC = nil
+	}
+}
+func (dt *DomainRoTx) valsCursor(tx kv.Tx) (c kv.Cursor, err error) {
+	eface := *(*[2]uintptr)(unsafe.Pointer(&tx))
+	if dt.valsC != nil { // run in assert mode only
+		fmt.Printf("cmp e1=%x e2=%x s=%x d=%s\n", eface[0], eface[1], dt.vcParentPtr.Load(), dt.d.filenameBase)
+		if !dt.vcParentPtr.CompareAndSwap(eface[1], eface[1]) { // cant swap when parent ptr is different
```
Comment (AskAlexSharov) [inline erigon-lib/state/domain.go:1721]: do we need atomic here? `sd` is not thread-safe

Comment (AskAlexSharov) [reply erigon-lib/state/domain.go:1721]: FYI: `tx` has method `tx.ViewID()`

**Comment by AskAlexSharov** [comment]
this branch, `ethmainnet` at startup got:
```
[INFO] [12-07|12:49:31.453] [4/6 Execution] Done                     blk=21231799 blks=168 blk/s=11080.0 txs=1 tx/s=65 gas/s=0 buf=940B/512.0MB stepsInDB=0.00 step=1683.0 inMem=false alloc=5.9GB sys=6.0GB
panic: tx passed into ShredDomains is immutable: cursor parent tx c17ad67360; current tx c17b728ac0

goroutine 20132 [running]:
github.com/erigontech/erigon-lib/state.(*DomainRoTx).valsCursor(0xc17b3b0700, {0x71d64df015d8, 0xc17b728ac0})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/state/domain.go:1731 +0x3c5
github.com/erigontech/erigon-lib/state.(*DomainRoTx).getLatestFromDb(0xc17b3b0700, {0x5b7f831, 0x1, 0x1}, {0x71d64df015d8?, 0xc17b728ac0?})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/state/domain.go:1751 +0x4d
github.com/erigontech/erigon-lib/state.(*DomainRoTx).GetLatest(0xc17b3b0700, {0x5b7f831, 0x1, 0x1}, {0x71d64df015d8?, 0xc17b728ac0?})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/state/domain.go:1808 +0x15e
github.com/erigontech/erigon-lib/state.(*AggregatorRoTx).GetLatest(0x44?, 0xf7c0?, {0x5b7f831?, 0x1c0?, 0x1?}, {0x71d64df015d8?, 0xc17b728ac0?})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/state/aggregator.go:1857 +0x45
github.com/erigontech/erigon-lib/state.(*SharedDomains).GetLatest(0xc171f567e0, 0x4, {0x5b7f831, 0x1, 0x1})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/state/domain_shared.go:972 +0x92
github.com/erigontech/erigon-lib/state.(*SharedDomains).DomainPut(0xc171f567e0, 0x4, {0x5b7f831, 0x1, 0x1}, {0x0, 0x0, 0x0}, {0xc17c51db40, 0x4, ...}, ...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/state/domain_shared.go:1009 +0x9f
github.com/erigontech/erigon/core/rawdb/rawtemporaldb.AppendReceipt({0x33562b0, 0xc171f567e0}, 0x71d6395c7200?, 0x0)
	github.com/erigontech/erigon/core/rawdb/rawtemporaldb/accessors_receipt.go:85 +0x103
github.com/erigontech/erigon/eth/stagedsync.(*serialExecutor).execute(0xc17b5f6908, {0x335e198, 0xc0012d9d10}, {0xc17b70a488, 0x51, 0x0?})
	github.com/erigontech/erigon/eth/stagedsync/exec3_serial.go:136 +0x152
github.com/erigontech/erigon/eth/stagedsync.ExecV3({_, _}, _, {_, _}, _, {{0x3375810, 0xc000be8150}, 0x20000000, {0x1, ...}, ...}, ...)
	github.com/erigontech/erigon/eth/stagedsync/exec3.go:590 +0x22fa
github.com/erigontech/erigon/eth/stagedsync.ExecBlockV3(_, {_, _}, {{_, _}, {_, _}, _}, _, {0x335e198, ...}, ...)
	github.com/erigontech/erigon/eth/stagedsync/stage_execute.go:162 +0x20f
github.com/erigontech/erigon/eth/stagedsync.SpawnExecuteBlocksStage(_, {_, _}, {{_, _}, {_, _}, _}, _, {0x335e198, ...}, ...)
	github.com/erigontech/erigon/eth/stagedsync/stage_execute.go:256 +0x10e
github.com/erigontech/erigon/eth/stagedsync.PipelineStages.func10(0x9?, 0x0?, {0x3357ca0?, 0xc174583ce0?}, {{0x0, 0x0}, {0x0, 0x0}, 0x0}, {0x3374208, ...})
	github.com/erigontech/erigon/eth/stagedsync/default_stages.go:238 +0xf0
github.com/erigontech/erigon/eth/stagedsync.(*Sync).runStage(0xc174583ce0, 0xc1749875e0, {0x3375810, 0xc000be8150}, {{0x0, 0x0}, {0x0, 0x0}, 0x0}, 0x1, ...)
	github.com/erigontech/erigon/eth/stagedsync/sync.go:528 +0x190
github.com/erigontech/erigon/eth/stagedsync.(*Sync).Run(0xc174583ce0, {0x3375810, 0xc000be8150}, {{0x0, 0x0}, {0x0, 0x0}, 0x0}, 0x0?, 0x1)
	github.com/erigontech/erigon/eth/stagedsync/sync.go:412 +0x2ad
github.com/erigontech/erigon/turbo/stages.ProcessFrozenBlocks({0x335e198, 0xc0012d9d10}, {0x3375810, 0xc000be8150}, {0x339b648, 0xc001600570}, 0xc174583ce0, 0x0)
	github.com/erigontech/erigon/turbo/stages/stageloop.go:153 +0x15b
github.com/erigontech/erigon/turbo/execution/eth1.(*EthereumExecutionModule).Start(0xc17495bcc0, {0x335e198, 0xc0012d9d10})
	github.com/erigontech/erigon/turbo/execution/eth1/ethereum_execution.go:320 +0x165
created by github.com/erigontech/erigon/eth.(*Ethereum).Start in goroutine 1
	github.com/erigontech/erigon/eth/backend.go:1584 +0xdff
exit status 2
```

**Comment by awskii** [comment]
Cursor keeps instance of `tx` inside of it. When `tx` is Committed/Rollback-ed, all bounded Cursors are invalidated. But existing `kv.Cursor` pointer (like one in `DomainRoTx`) does not know underlying `tx` has been expired. So regardless to check of ViewID we get NPE during `mdbx.(*Cursor).getVal` because mdbxCurosr (`c._c`) is invalid. 
Added checks with panic, running on mainnet to verify.

**Comment by awskii** [comment]
Previous panic you had was realted to issue that sometimes we use tx, sometimes pass SharedDomains as kv.Tx implementation. Previous idea was to check pointers to given tx implementations and if they're mismatching, panic. So they did: we queried `AggregatorRoTx` with both tx and SharedDomains causing ptr mismatch => panic

**Comment by awskii** [comment]
For the check being active, hid it behind "ERIGON_AGG_ASSERTS=true" env

**Pull request merged**